icx-http-server/src/main.rs

```diff
@@ -0,0 +1,272 @@
+    }
+
+    // Look into the request by header.
+    if let Some(referer_header) = request.headers().get("referer") {
+        if let Ok(referer) = referer_header.to_str() {
+            if let Ok(referer_uri) = hyper::Uri::from_str(referer) {
+                if let Some(canister_id) = resolve_canister_id_from_query(&referer_uri) {
+                    return Some(canister_id);
+                }
+            }
+        }
+    }
+
+    None
+}
+
+async fn forward_request(
+    request: Request<Body>,
+    agent: Arc<Agent>,
+) -> Result<Response<Body>, Box<dyn Error>> {
+    let canister_id = match resolve_canister_id(&request) {
+        None => {
+            return Ok(Response::builder()
+                .status(StatusCode::BAD_REQUEST)
+                .body("Could not find a canister id to forward to.".into())
+                .unwrap())
+        }
+        Some(x) => x,
+    };
+
+    let method = request.method().to_string();
+    let uri = request.uri().clone();
+    let headers = request
+        .headers()
+        .into_iter()
+        .filter_map(|(name, value)| {
+            Some(HeaderField(
+                name.to_string(),
+                value.to_str().ok()?.to_string(),
+            ))
+        })
+        .collect();
+
+    let entire_body = body::to_bytes(request.into_body()).await?.to_vec();
+
+    HttpRequestCanister::create(agent.as_ref(), canister_id)
+        .http_request(method, uri.to_string(), headers, &entire_body)
+        .call()
+        .await
+        .map_err(Into::<Box<dyn Error>>::into)
+        .and_then(|(http_response,)| {
+            let mut builder =
+                Response::builder().status(StatusCode::from_u16(http_response.status_code)?);
+            for HeaderField(name, value) in http_response.headers {
+                builder = builder.header(&name, value);
+            }
+            Ok(builder.body(http_response.body.into())?)
+        })
+}
+
+fn is_hop_header(name: &str) -> bool {
+    name.to_ascii_lowercase() == "connection"
+        || name.to_ascii_lowercase() == "keep-alive"
+        || name.to_ascii_lowercase() == "proxy-authenticate"
+        || name.to_ascii_lowercase() == "proxy-authorization"
+        || name.to_ascii_lowercase() == "te"
+        || name.to_ascii_lowercase() == "trailers"
+        || name.to_ascii_lowercase() == "transfer-encoding"
+        || name.to_ascii_lowercase() == "upgrade"
+}
+
+/// Returns a clone of the headers without the [hop-by-hop headers].
+///
+/// [hop-by-hop headers]: http://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html
+fn remove_hop_headers(
+    headers: &hyper::header::HeaderMap<hyper::header::HeaderValue>,
+) -> hyper::header::HeaderMap<hyper::header::HeaderValue> {
+    let mut result = hyper::HeaderMap::new();
+    for (k, v) in headers.iter() {
+        if !is_hop_header(k.as_str()) {
+            result.insert(k.clone(), v.clone());
+        }
+    }
+    result
+}
+use hyper::http::uri::Parts;
```

Comment:
Should this be at the top of the file with the other `use` statements?

Reply:
Yes, I was experimenting for a while.